### PR TITLE
wallet: unify max signature logic

### PR DESCRIPTION
**Pull request by S3RK**
Currently `DummySignTx` and `DummySignInput` use different ways to determine signature size.
This PR unifies the way wallet estimates signature size for various inputs. 
Instead of passing boolean flags from calling code the `use_max_sig` is now calculated at the place of signature creation using information available in `CCoinControl`

### CODE REVIEW

File: src/wallet/spend.cpp [hunk 1]
```diff
@@ -198,7 +193,7 @@ CoinsResult AvailableCoins(const CWallet& wallet,
             // Filter by spendable outputs only
             if (!spendable && only_spendable) continue;
 
-            int input_bytes = GetTxSpendSize(wallet, wtx, i, (coinControl && coinControl->fAllowWatchOnly));
+            int input_bytes = CalculateMaximumSignedInputSize(output, &wallet, false);
```
Comment (furszy) [inline src/wallet/spend.cpp]: why not `use_max_sig=coinControl && coinControl->fAllowWatchOnly` ?

File: src/wallet/spend.cpp [hunk 2]
```diff
@@ -289,8 +284,9 @@ std::map<CTxDestination, std::vector<COutput>> ListCoins(const CWallet& wallet)
             ) {
                 CTxDestination address;
                 if (ExtractDestination(FindNonChangeParentOutput(wallet, *wtx.tx, output.n).scriptPubKey, address)) {
+                    const auto out = wtx.tx->vout.at(output.n);
                     result[address].emplace_back(
-                        COutPoint(wtx.GetHash(), output.n), wtx.tx->vout.at(output.n), depth, GetTxSpendSize(wallet, wtx, output.n), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ false, wtx.GetTxTime(), CachedTxIsFromMe(wallet, wtx, ISMINE_ALL));
+                            COutPoint(wtx.GetHash(), output.n), out, depth, CalculateMaximumSignedInputSize(out, &wallet, false), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ false, wtx.GetTxTime(), CachedTxIsFromMe(wallet, wtx, ISMINE_ALL));
```
Comment (furszy) [inline src/wallet/spend.cpp]: same as above, let's document the `use_max_sig` flag.

File: src/wallet/test/coinselector_tests.cpp [hunk 3]
```diff
@@ -86,7 +86,8 @@ static void add_coin(std::vector<COutput>& coins, CWallet& wallet, const CAmount
     auto ret = wallet.mapWallet.emplace(std::piecewise_construct, std::forward_as_tuple(txid), std::forward_as_tuple(MakeTransactionRef(std::move(tx)), TxStateInactive{}));
     assert(ret.second);
     CWalletTx& wtx = (*ret.first).second;
-    coins.emplace_back(COutPoint(wtx.GetHash(), nInput), wtx.tx->vout.at(nInput), nAge, GetTxSpendSize(wallet, wtx, nInput), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, wtx.GetTxTime(), fIsFromMe, feerate);
+    const auto txout = wtx.tx->vout.at(nInput);
```
Comment (furszy) [inline src/wallet/test/coinselector_tests.cpp:89]:
```suggestion
    const auto& output = wtx.tx->vout.at(nInput);
```

File: src/wallet/test/coinselector_tests.cpp [hunk 4]
```diff
@@ -86,7 +86,8 @@ static void add_coin(std::vector<COutput>& coins, CWallet& wallet, const CAmount
...
+    coins.emplace_back(COutPoint(wtx.GetHash(), nInput), txout, nAge, CalculateMaximumSignedInputSize(txout, &wallet, false), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, wtx.GetTxTime(), fIsFromMe, feerate);
```
Comment (furszy) [inline src/wallet/test/coinselector_tests.cpp]:
```suggestion
    coins.emplace_back(COutPoint(wtx.GetHash(), nInput), txout, nAge, CalculateMaximumSignedInputSize(txout, &wallet, false), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, wtx.GetTxTime(), fIsFromMe, feerate);
```

File: src/bench/coin_selection.cpp [hunk 5]
```diff
@@ -56,10 +56,11 @@ static void CoinSelection(benchmark::Bench& bench)
     addCoin(3 * COIN, wallet, wtxs);
 
     // Create coins
+    CCoinControl coin_control:
```
Comment (furszy) [inline src/bench/coin_selection.cpp:59]:
```suggestion
    wallet::CCoinControl coin_control;
```

**Review by furszy — COMMENTED** [review]
Aside from the quick code review that made, want to point you in #24699 direction.

The first two commits there are implementing part of what you did here. Would be nice to get that one reviewed first.

File: src/wallet/spend.cpp (hunk 1 above)
Comment (S3RK) [reply src/wallet/spend.cpp]: I'd like to remove the flag

File: src/wallet/spend.cpp (hunk 2 above)
Comment (S3RK) [reply src/wallet/spend.cpp]: It's removed in the next commit

File: src/wallet/spend.cpp [hunk 6]
```diff
@@ -447,7 +443,7 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
             if (ptr_wtx->tx->vout.size() <= outpoint.n) {
                 return std::nullopt;
             }
-            input_bytes = GetTxSpendSize(wallet, *ptr_wtx, outpoint.n, false);
+            input_bytes = CalculateMaximumSignedInputSize(ptr_wtx->tx->vout[outpoint.n], &wallet, false);
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp]: Same as above

File: src/wallet/test/coinselector_tests.cpp (hunk 4 above)
Comment (S3RK) [reply src/wallet/test/coinselector_tests.cpp]: Can't spot the difference

Comment (furszy) [reply src/wallet/test/coinselector_tests.cpp]: It it the same as this one https://github.com/bitcoin/bitcoin/pull/25481#discussion_r907362233

File: src/bench/coin_selection.cpp (hunk 5 above)
Comment (achow101) [reply src/bench/coin_selection.cpp:59]: Or pass a nullptr instead of an empty `CCoinControl`.

Comment (S3RK) [reply src/bench/coin_selection.cpp]: Done. Replaced with nullptr

**Comment by S3RK** [comment]
@furszy thanks for the review and the pointer. I don't want to add hints for the flag, because it's removed in the next commit. 
I made this PR compatible with #24699 and now they nicely compliment each other. 
008d1178c8b09aa9190b188f399fc0f5818c7499 Is fully covered by the first commit here and it does even more.
c685bf545416abe719d2821e8156ff1d0bb2d749 Is taking a slightly different direction, but it's compatible with the code here
It seems either PR can go first.

**Comment by achow101** [comment]
ACK d54c5c8b1b1a38b5b38e6878aea0fa8d6c1ad7e9

**Review by theStack — APPROVED** [review]
Code-review ACK d54c5c8b1b1a38b5b38e6878aea0fa8d6c1ad7e9

**Review by murchandamus — COMMENTED** [review]
Post-merge ACK d54c5c8b1b1a38b5b38e6878aea0fa8d6c1ad7e9

**Pull request merged**